lib/tasks_private/lib/aws_instance_data_parser.rb

```diff
@@ -0,0 +1,316 @@
+  CLUSTERABLE_TYPES = %w(m4 c3 c4 cr1 r4 x1 hs1 i2 g2 p2 d2).freeze
+
+  # some types data missing this
+  EBS_OPTIMIZED_TYPES = %w(x1).freeze
+
+  # some types data missing 'intel' features info
+  CPU_FEATURES = {
+    'Intel Xeon E5-2670'                              => %i(avx turbo).freeze,
+    'Intel Xeon E5-2670 v2 (Ivy Bridge/Sandy Bridge)' => %i(avx turbo).freeze,
+    'Intel Xeon E5-2666 v3 (Haswell)'                 => %i(avx avx2 turbo).freeze,
+    'Intel Xeon E5-2686 v4 (Broadwell)'               => %i(avx avx2 turbo).freeze,
+  }.freeze
+
+  # considering all types have it except of legacy ones
+  NON_AES_NI_TYPES = %w(t1 m1).freeze
+
+  # some types data missing this (here is only those who missing)
+  # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/
+  #   enhanced-networking.html#supported_instances
+  ENHANCED_NETWORKING_TYPES = %w(f1).freeze
+
+  ParsedName = Struct.new(:base_type, :size_factor, :size_name)
+  ParsedStorage = Struct.new(:volumes, :size, :type)
+
+  private_constant(*constants(false).without(:REQUIRED_ATTRIBUTES))
+
+  attr_reader :product_data
+
+  def initialize(product_data)
+    @parsed = false
+    @product_data = product_data
+    @unknown_values = {}
+  end
+
+  def result
+    [instance_data, unknown_values]
+  end
+
+  def instance_data
+    parse! unless @parsed
+    @instance_data
+  end
+
+  def unknown_values
+    parse! unless @parsed
+    @unknown_values
+  end
+
+  ### individual attributes
+
+  ## general
+
+  def current_generation?
+    product_data['currentGeneration'] == 'Yes'
+  end
+
+  def current_version?
+    !!product_data['currentVersion']
+  end
+
+  def instance_type
+    product_data['instanceType']
+  end
+
+  def instance_family
+    product_data['instanceFamily']
+  end
+
+  delegate :base_type, :to => :parsed_name
+
+  def description
+    description = [base_type.upcase]
+    description << instance_family unless POPULAR_TYPES.include?(base_type)
+    description << (size_name == 'xlarge' ? "#{size_factor}XL" : size_name.capitalize)
+    description.join(' ')
+  end
+
+  ## virtualization
+
+  def virtualization_type
+    VIRT_TYPES[base_type]
+  end
+
+  delegate :size_factor, :size_name, :to => :parsed_name
+
+  def vcpus
+    product_data['vcpu'].to_i
+  end
+
+  ## cpu
+
+  def physical_processor
+    product_data['physicalProcessor']
+  end
+
+  def cpu_clock_speed
+    product_data['clockSpeed']
+  end
+
+  def cpu_arches
+    CPU_ARCHES.fetch(product_data['processorArchitecture']) do |cpu_arch|
+      save_unknown(cpu_arch)
+    end
+  end
+
+  def processor_features
+    product_data['processorFeatures']
+  end
+
+  def intel_avx?
+    product_data['intelAvxAvailable'] == 'Yes' ||
+      !!(processor_features =~ INTEL_AVX_REGEXP) ||
+      CPU_FEATURES[physical_processor]&.include?(:avx)
+  end
+
+  def intel_avx2?
+    product_data['intelAvx2Available'] == 'Yes' ||
+      !!(processor_features =~ INTEL_AVX2_REGEXP) ||
```

Comment:
Instead of `!!(foo =~ REGEXP)`, you can also do `REGEXP.match?(foo)`, which reads a little nicer.

Reply:
`Regexp` doesn't have such method:
```
//.methods.grep /match/
=> [:match]
```